cpp/include/cudf/utilities/error.hpp

```diff
@@ -29,13 +31,35 @@ namespace cudf {
  * @file
  */
 
+/**
+ * @brief The struct to store the current stacktrace upon its construction.
+ */
+struct stacktrace_recorder {
+  stacktrace_recorder()
+    // Exclude the current stackframe, as it is this constructor.
+    : _stacktrace{cudf::detail::get_stacktrace(cudf::detail::capture_last_stackframe::NO)}
+  {
+  }
+
+ public:
+  /**
+   * @brief Get the stored stacktrace captured during object construction
+   *
+   * @return The pointer to a null-terminated string storing the output stacktrace
+   */
+  char const* stacktrace() const { return _stacktrace.c_str(); }
+
+ protected:
+  std::string const _stacktrace;  //!< The whole stacktrace stored as one string.
+};
+
 /**
  * @brief Exception thrown when logical precondition is violated.
  *
  * This exception should not be thrown directly and is instead thrown by the
  * CUDF_EXPECTS macro.
  */
-struct logic_error : public std::logic_error {
```

Comment:
Folks should be aware that this is changing the size of the exception type even when compiling in non-stacktrace mode.

That isn't to say it's a good idea to _not_ inherit from `stacktrace_recorder` in non-stacktrace mode because there would be the possibility for weird ABI issues with different builds of libcudf having different sizes for exception types. 

The ideal way to do this without changing the size of the exception type would be to include the stack trace in the `what()` string such that there is only a single `string` member.

Reply:
The "what()" string is accessed for logging. If we extend it, it may be very large for downstream applications to print out every time. This issue has been mentioned before, please see https://github.com/rapidsai/cudf/issues/12422#issuecomment-1371387818.

Reply:
There is another solution for this concern: Only modify the exception structs with compiler check `#ifdef CUDF_BUILD_STACKTRACE_DEBUG`. The `char* stacktrace()` will always be an empty string if that option is not defined. How do you think about it?

Reply:
Does the recorder need to be part of the signature? 
The stack trace does not seem to have any real association with the exception object.
Could it not just be free function used like this?

```
try {
  // cudf API calls
} catch (cudf::logic_error const& e) {
  std::cout << e.what() << std::endl;
  std::cout << stacktrace() << std::endl;
  throw e;
} 
```

Reply:
Unfortunately not. `stacktrace()` gives you the call stack up to the point inside `stacktrace()`. From there, you cannot trace back to where the exception was created.

Reply:
An alternative would be for each exception type to own a `stacktrace_recorder` that is constructed in its constructor. That approach would require ignoring 2 frames instead of just 1, though, and it would be more verbose. I don't have a strong opinion that that is a necessary change here.